Create executable functions from Sympy expressions.

mdewing opened the pull request:
Use llvmlite to JIT compile Sympy expressions into executable code.
The initial implementation only works with scalar expressions.
These expressions can contain function calls to single-argument math functions
(such as 'exp', 'sin', etc.)

mdewing committed:
Create executable functions from Sympy expressions.

Use llvmlite to JIT compile Sympy expressions into executable code.
The initial implementation only works with scalar expressions.
These expressions can contain function calls to single-argument math functions
(such as 'exp', 'sin', etc.)

moorepants commented:
Very cool! Thanks for the addition. I'll add some inline comments.

moorepants commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,175 @@
+from sympy.external import import_module
+
+ll = import_module('llvmlite.ir').ir
+llvm = import_module('llvmlite.binding').binding
```
There should be a call (probably in the `__init__` method of the class) that checks for this optional dependency and raises and error if it isn't installed, for example:

```python
if ll is None:
    raise ...
```

moorepants commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,175 @@
+llvm.initialize_native_asmprinter()
+
+
+class LLVMJitPrinter(sympy.printing.printer.Printer):
```
This is minor, but the typical style in SymPy code is to have `from sympy.printing.printer import Printer` as the import and then just subclass `Printer`.

moorepants commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,175 @@
+
+
+class LLVMJitPrinter(sympy.printing.printer.Printer):
+    def __init__(self, *args, **kwargs):
```
This class and `__init__` method needs a docstring. Note that we follow the numpydoc style.

moorepants commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,175 @@
+current_link_suffix = 0
+
+
+def llvm_jit_code(expr, args=None, link_name=None):
```
This needs a docstring that follows numpydoc style.

moorepants commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,175 @@
+
+
+def get_jit_callable(expr, args=None, link_name=None):
+    '''Create an executable function from a Sympy expression'''
```
Please explain the function arguments in the docstring.

moorepants commented on sympy/printing/tests/test_llvmjit.py:
```diff
@@ -0,0 +1,45 @@
+if import_module('llvmlite'):
+    import sympy.printing.llvmjitcode as g
+else:
+    disabled = True
```
What's this for?

markdewing replied:
I don't know - I was following the theano printer as an example, and that test includes this bit of code.   It appears to be checked in conftest.py ,in check_disabled.  I assume this is to disable the tests if the dependency is not installed (?)

moorepants replied:
I just didn't see the use of disabled below, so I don't think it actually disables the tests.

asmeurer replied:
Maybe it's for py.test

moorepants replied:
I see. That seems like a bad design. So, if you use py.test you can't use a variable called `disabled` except for signaling py.test to run? I haven't found anything in the py.test docs about this yet.

asmeurer replied:
Actually it looks like it's for the SymPy test runner. I don't know if this signals py.test as well (I didn't see anything in the py.test docs, but I didn't look to hard either).

moorepants commented:
These tests will fail on Travis due to the fact that llvmlite isn't installed. It would be great if we could add in that dependency to at least one of the test calls so that this code won't get bit rot. I'd like to see that added in this PR. The other option would be to disable the tests if llvmlite isn't installed.

moorepants commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,175 @@
+
+Scalar expression
+-----------------
+  import sympy.printing.llvmlitecode as g
```
lit or jit?

moorepants replied:
Oh, "lite"...

moorepants commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,175 @@
+  a = Symbol('a')
+  e = a*a + a + 1
+  e1 = g.get_jit_callable(e, args=[a])
+  print(e1(1.1), e.subs('a',1.1))
```
This should follow our docstring style so that it get's tested, i.e. use `>>>` to proceed input lines.

mdewing committed:
JIT compile of Sympy expressions with llvmlite

If llvmlite is not installed, it should import cleanly, but then raise
an exception when 'get_jit_callable' is called.
Improved documentation.  Example is now in doctest format.

Only tested on 64-bit linux.

markdewing commented:
Thanks for the comments.    Branch is updated.
Should I squash the two commits into one for a cleaner history?

moorepants commented:
Squashing is not necessary.

cbm755 commented:
Perhaps I'm missing the point and have not looked carefully, but could this be merged into `codegen`?

mdewing committed:
Fix failing doctests for llvmlite JIT

Add dependency on llvmlite module for doctests.

markdewing commented:
This interface and purpose for 'get_jit_callable' is similar to autowrap in codegen.  Maybe it would be better to add this as a new language and backend under autowrap? (language="JIT" , backend="lllvmlite"?)

moorepants commented:
For all other cases (except theano?) we separate the printing from the compiling/wrapping. I think it would be nicer if we could add in this backend to autowrap. But we could get this merged as is and then open an issue for adding both the theano and llvm backends to autowrap.

asmeurer commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,192 @@
+from sympy.utilities.decorator import doctest_depends_on
+
+llvmlite = import_module('llvmlite')
+if llvmlite:
```
If this is false you will just get NameErrors in the code.

moorepants replied:
Yes, that is what happens. You need to raise an import error in the Printer's __init__ function based on `if llvmlite`.

asmeurer commented:
I agree that llvm_jit_code should be callable from autowrap (it can also stay as its own function in this module as well). Shouldn't it be `language='LLVM'` instead of `language='JIT'`, though?

asmeurer commented:
Regarding Travis, I have enable conda for one of the build jobs at https://github.com/sympy/sympy/pull/10084. Once it is merged we can add llvmlite to it for this PR. Please *don't* add new builds to the matrix.

markdewing commented:
With autowrap, using `language='LLVM'` (and ignore `backend`) would be better.

moorepants commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,192 @@
+    llvm.initialize_native_asmprinter()
+
+
+class LLVMJitPrinter(Printer):
```
I don't really know how all the LLVM stuff works, but a printer should be able to print something. Can this print any actual code? Does it generate assembly code? What are the instructions that are passed to LLVM?

asmeurer replied:
This matches what the theano "printers" do.

moorepants replied:
The theano printer has a `doprint` method. See:

```
Python 3.5.1 |Continuum Analytics, Inc.| (default, Dec  7 2015, 11:16:01) 
Type "copyright", "credits" or "license" for more information.

IPython 4.0.1 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: from sympy import *

In [2]: from sympy.abc import *

In [3]: from sympy.printing.theanocode import the
theano           theano_code      theano_function  

In [3]: from sympy.printing.theanocode import TheanoPrinter

In [4]: p = TheanoPrinter()

In [5]: p.doprint(a**2 + b**2)
Out[5]: Elemwise{add,no_inplace}.0
```

But maybe LLVM doesn't have a string rep.

markdewing replied:
The LLVM IR representation can be printed (it looks a bit like assembly code).  It might be useful to have a function that returns the IR.  What I don't know yet is if fragments of LLVM IR can be produced, or if the entire function needs to be produced at once.  The current code printers produce single C or Fortran expressions, and the codgen code puts them in functions.  I don't know if the split is possible or reasonable for creating LLVM IR.

moorepants replied:
Ok. If it is possible and not too hard, I would be in favor of making that happen. But if not, this is fine.

asmeurer replied:
That's not a string, though

```
In [9]: type(p.doprint(a**2 + b**2))
Out[9]: theano.tensor.var.TensorVariable
```

asmeurer replied:
Does llvmlite have a function to spit out the IR? If so, one could just use that (and we could document it). If not, then it's not worth it. The point of llvmlite is to have an abstraction over the IR. Writing a printer directly to it would defeat the purpose, and be a lot more work (you'd have to essentially rewrite llvmlite I'd guess).

moorepants replied:
That's fine. No need to do lots of extra work, but if it happens that we could fit this into the printer "mold" more precisely with little work then I think we should.

markdewing replied:
The IR is converted to a string before assembly (the line `strmode = str(module)` in `llvm_jit_code`).  The question is how to give users access.

The `llvm_jit_code` function contains a lot of LLVM setup and management code.  My preference would be to keep this code out of the LLVMJITPrinter class (as future features are going to increase the size of this class anyway). 
Perhaps the `llvm_jit_code` should be converted into a class, with smaller methods, and would allow access to the string representation of the IR.

moorepants commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,192 @@
+    for i, a in enumerate(func_args):
+        name = str(a)
+        fn.args[i].name = name
+        lj.param_dict[name] = fn.args[i]
```
Hashing on the string version of the SymPy type can cause problems. For example, two symbols can have the same string rep but different assumptions. I think your hash key should be the SymPy object itself.

asmeurer replied:
But at the LLVM level aren't things only identifiable by their string value?

markdewing replied:
This particular dict is for mapping the Sympy symbol in the expression to the LLVM function argument (accessed by indexed position - `fn.args[i]`) for proper replacement while building the LLVM tree.  I can change it.

moorepants commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,192 @@
+    lj.builder.ret(ret)
+
+    strmod = str(module)
+    if False:
```
This should be made a kwarg or dropped.

moorepants commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,192 @@
+
+    fptr = exe_eng.get_pointer_to_function(llmod.get_function(link_name))
+
+    return fptr
```
To follow the pattern of other Printers, this should return something that is printable in ASCII code. A lot of what is in this function seems more like what should be in the "callable" function.

moorepants commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,192 @@
+
+
+@doctest_depends_on(modules=('llvmlite',))
+def get_jit_callable(expr, func_args=None):
```
To match the theano stuff, would `jit_function` or `llvm_function` be a better name? I'm not a big fan of calling it `get_` anything.

asmeurer replied:
`get_llvm_callable` seems like a better name. LLVM and JIT are not interchangeable. There are other types of JIT than LLVM.

asmeurer replied:
Or even just `llvm_callable`. "get" is redundant in a function name.

moorepants commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,192 @@
+        return ll.Constant(self.fp_type, float(n))
+
+    def _print_Integer(self, expr):
+        return ll.Constant(self.fp_type, float(expr.p))
```
Why is everything turned into a double?

markdewing replied:
For now, the assumption is that all computations will be double precision floating point.  This will probably need to change in the future (some integer values may be array bounds or array indices, or users may want to switch to single precision, etc.).
Since LLVM is low-level, values need to be explicitly declared as the right type (or cast to the right type)

moorepants replied:
Ok, it would be worth adding that to the docstring for users to know. I'm also curious what happens if there are SymPy types that don't convert. Is it possible to give a nice error message to users?

mdewing committed:
Cleanup for LLVM JIT.

Change name of primary function to 'llvm_callable'.
Raise exception when attempting to convert unsupported Sympy nodes.
Pass important class members as constructor arguments.

mdewing committed:
Split large LLVM JIT function into smaller methods

Create a class to organize the pieces of creating an LLVM JIT function.

mdewing committed:
Handle two calls to the same function in LLVM JIT

Create a cache for wrappers that call external math functions.

markdewing commented:
Is there anything more I need to do on this PR?
I've got some more additions, like extending it to handle sums over numpy arrays and making a JIT compiled callback for scipy integration routines.  I can add them to this PR, but they might be better as a separate PR

asmeurer commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,235 @@
+
+
+@doctest_depends_on(modules=('llvmlite',))
+def llvm_callable(expr, func_args=None):
```
This fails if you don't pass the keyword argument.

asmeurer replied:
Any reason to not make it required? Maybe it should be `llvm_callable(args, expr)` like lambdify.

asmeurer commented on sympy/printing/llvmjitcode.py:
```diff
@@ -0,0 +1,235 @@
+    >>> from sympy.abc import a
+    >>> e = a*a + a + 1
+    >>> e1 = jit.llvm_callable(e, func_args=[a])
+    >>> e.subs('a',1.1)   # Evaluate via substitution
```
Don't use strings.

asmeurer commented:
Can you add something to the codegen docs about this? I'm assuming `llvm_callable` is intended to be the primary entry point for this module.

moorepants commented:
One last comment from me. Think about making some of your methods private (use a leading underscore in the name). If you make them all public, we need full docstrings that follow the numpydoc convention along with examples. We also have to make the public ones backwards compatible in future releases, so if you carefully choose what should be public it will make things easier to change in the future.

After that, this looks good. +1.

Thanks for this cool addition!

mdewing committed:
Update LLVM JIT arguments.

Change the order of the arguments to llvm_callable to match lambdify.
Also mark some of the functions as private.  For now, only the
llvm_callable function is intended to be used outside the module.

markdewing commented:
Yes, `llvm_callable` is the main entry point, and the only one intended to be public.
The latest update has all changes except for adding documentation.

asmeurer commented:
+1

asmeurer commented:
Oh, actually, one more change. You may need to merge in master first, but add `llvmlite` to the `conda install` list in `.travis.yml` (in the AUTOWRAP build) so that this actually gets tested on Travis.

mdewing committed:
Merge branch 'master' into jit

mdewing committed:
Add llvmlite to autowrap section for testing JIT.

mdewing committed:
Remove comment causing error in .travis.yml.

asmeurer commented:
Add llvmlite to the conda line that is already there.

mdewing committed:
Add llvmlite to list of packages for conda install
